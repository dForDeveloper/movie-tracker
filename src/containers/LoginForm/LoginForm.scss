.LoginForm {
  width: 30%;
  margin: 30px auto;
  display: flex;
  padding: 25px 20px;
  flex-direction: column;
  font-family: $body-font;
  background-color: $opaque-white;
  border: 6px solid $border-color;
  box-shadow: 12px 12px rgba(243, 241, 255, 0.25);
}

.login-form-title {
  text-align: center;
  margin-bottom: 15px;
  font-family: $heading-font;
  font-size: 40px;
  text-shadow: 2px 2px $off-white;
}

.login-button {
  width: 60%;
  margin: 15px auto 0px auto;
  padding: 8px;
  font-size: 22px;
  text-transform: capitalize;
  font-weight: bold;
  border: 3px solid $border-color;
  border-radius: 5px;
  background-color: $steel-grey;
  color: $off-white;
  letter-spacing: 1px;
}

.login-button:hover {
  background-color: $powder-blue;
  color: $text-color;
  cursor: pointer;
}