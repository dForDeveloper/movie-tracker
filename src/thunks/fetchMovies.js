import { addMovies } from '../actions';
import { fetchData } from '../utils/api';

export const fetchMovies = (url) => {
  return async (dispatch) => {
    const data = await fetchData(url);
    dispatch(addMovies(data.results));
  }
}