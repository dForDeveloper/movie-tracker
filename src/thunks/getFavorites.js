import { setFavorites } from '../actions';
import { fetchData } from '../utils/api';

export const getFavorites = (userID) => {
  return async (dispatch) => {
    const favesUrl = `http://localhost:3000/api/users/${userID}/favorites`;
    const response = await fetchData(favesUrl);
    const favorites = response.data.map(favorite => favorite.movie_id);
    dispatch(setFavorites(favorites));
  }
}