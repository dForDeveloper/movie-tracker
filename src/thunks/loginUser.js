import { fetchData } from '../utils/api';
import { setUser, toggleLoginPrompt } from '../actions';
import { getFavorites } from './getFavorites';

export const loginUser = (email, password) => {
  return async (dispatch) => {
    try {
      const loginUrl = 'http://localhost:3000/api/users';
      const options = {
        method: 'POST',
        body: JSON.stringify({ email, password }),
        headers: {
          'Content-Type': 'application/json'
        }
      };
      const response = await fetchData(loginUrl, options);
      const { name, id } = response.data;
      dispatch(setUser({ name, id }));
      await dispatch(getFavorites(id));
      dispatch(toggleLoginPrompt(false));
      localStorage.setItem('user', JSON.stringify({ name, id }));
      return response.status;
    } catch {
      return 'error';
    }
  }
}