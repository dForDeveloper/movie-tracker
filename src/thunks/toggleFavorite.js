import { getFavorites } from './getFavorites';
import { fetchData } from '../utils/api';

export const toggleFavorite = (movie, isFavorite) => {
  return async (dispatch) => {
    const { user_id, movie_id } = movie;
    if (isFavorite) {
      const url = `http://localhost:3000/api/users/${user_id}/favorites/${movie_id}`;
      const options = {
        method: 'DELETE',
        body: JSON.stringify({ user_id, movie_id }),
        headers: {
          'Content-Type': 'application/json'
        }
      }
      await fetchData(url, options);
    } else {
      const url = 'http://localhost:3000/api/users/favorites/new';
      const options = {
        method: 'POST',
        body: JSON.stringify(movie),
        headers: {
          'Content-Type': 'application/json'
        }
      };
      await fetchData(url, options);
    }
    dispatch(getFavorites(user_id));
  }
}