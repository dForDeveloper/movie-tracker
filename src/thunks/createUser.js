import { fetchData } from '../utils/api';
import { loginUser } from './loginUser';

export const createUser = (name, email, password) => {
  return async (dispatch) => {
    try {
      const url = 'http://localhost:3000/api/users/new';
      const options = {
        method: 'POST',
        body: JSON.stringify({ name, email, password }),
        headers: {
          'Content-Type': 'application/json'
        }
      }
      const response = await fetchData(url, options);
      await dispatch(loginUser(email, password));
      return response.status;
    } catch {
      return 'error';
    }
  }
}