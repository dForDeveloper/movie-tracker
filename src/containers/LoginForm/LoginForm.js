import React, { Component } from 'react';
import { connect } from 'react-redux';
import { Redirect } from 'react-router-dom';
import PropTypes from 'prop-types';
import { loginUser } from '../../thunks/loginUser';

export class LoginForm extends Component {
  constructor() {
    super();
    this.state = {
      email: '',
      password: '',
      status: ''
    }
  }

  handleChange = (event) => {
    const { id, value } = event.target;
    this.setState({ [id]: value });
  }

  handleSubmit = async (event) => {
    event.preventDefault();
    const { email, password } = this.state;
    const status = await this.props.loginUser(email, password);
    this.setState({ status });
  }

  render() {
    let { status } = this.state;
    return (
      <form onSubmit={this.handleSubmit} className="LoginForm">
        <h2>Log In</h2>
        <label htmlFor="email">Email:</label>
        <input type="email" id="email" className="email-input" onChange={this.handleChange} />
        <label htmlFor="password">Password:</label>
        <input type="password" id="password" onChange={this.handleChange} />
        <input type="submit" value="login" />
        {status === 'error' && <p>Email and password do not match.</p>}
        {status === 'success' && <Redirect to='/' />}
      </form>
    );
  }
}

export const mapDispatchToProps = (dispatch) => ({
  loginUser: (email, password) => dispatch(loginUser(email, password))
});

export default connect(null, mapDispatchToProps)(LoginForm);

LoginForm.propTypes = {
  loginUser: PropTypes.func,
  history: PropTypes.object,
  location: PropTypes.object,
  match: PropTypes.object
};