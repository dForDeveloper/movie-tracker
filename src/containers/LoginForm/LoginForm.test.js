import React from 'react';
import { shallow } from 'enzyme';
import { LoginForm, mapDispatchToProps } from './LoginForm';
import { loginUser } from '../../thunks/loginUser';

const mockProps = {
  loginUser: jest.fn(),
  history: {},
  location: {},
  match: {}
}

describe('LoginForm', () => {
  let wrapper;
  beforeEach(() => {
    wrapper = shallow(
      <LoginForm {...mockProps} />
    );
  });

  describe('LoginForm container', () => {
    it('should properly render the component elements', () => {
      expect(wrapper).toMatchSnapshot();
    });
  
    it('should have the proper default state', () => {
      expect(wrapper.state()).toEqual({
        email: '',
        password: '',
        status: ''
      });
    });
  
    it('should set state based on an input field when text is typed', () => {
      const event = {target: {id: 'email', value: 'shannon'}};
      wrapper.find('#email').simulate('change', event);
      expect(wrapper.state('email')).toEqual('shannon');
    });
  });

  describe('mapDispatchToProps', () => {
    it.skip('should call dispatch when loginUser is called', () => {
      let dispatchMock = jest.fn();
      const expected = loginUser('email@test.io', 'shannon');
      const result = mapDispatchToProps(dispatchMock);
      result.loginUser('email@test.io', 'shannon');
      expect(dispatchMock).toHaveBeenCalledWith(expected);
    });
  });
  
});