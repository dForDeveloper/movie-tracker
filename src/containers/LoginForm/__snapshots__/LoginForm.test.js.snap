// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginForm LoginForm container should properly render the component elements 1`] = `
<form
  className="LoginForm"
  onSubmit={[Function]}
>
  <h2>
    Log In
  </h2>
  <label
    htmlFor="email"
  >
    Email:
  </label>
  <input
    className="email-input"
    id="email"
    onChange={[Function]}
    type="email"
  />
  <label
    htmlFor="password"
  >
    Password:
  </label>
  <input
    id="password"
    onChange={[Function]}
    type="password"
  />
  <input
    type="submit"
    value="login"
  />
</form>
`;
